File: Final_450/LinkedList/ImpleDLL.java
public class ImpleDLL {
    {
    static Node head = null;

    static class Node {
        int data;
        Node next;
        Node prev;

        Node(int d) {
            data = d;
            next = null;
            prev = null;
        }
    }

    public void append(Node data) {
        Node cur = head;
        if (head == null) {
            head = data;
            head.next = null;
            head.prev = null;
        } else {
            while (cur.next != null) {
                cur = cur.next;
            }
            cur.next = data;
            data.prev = cur;
            data.next = null;
        }
    }

    public void push(Node data) {
        if (head == null) {
            head = data;
            head.next = null;
            head.prev = null;
        }
        // Node cur = head;
        data.next = head;
        head.prev = data;
        data.prev = null;
        head = data;
    }

    public void printDll(Node head) {
        while (head != null) {
            System.out.print(head.data + "<-->");
            head = head.next;
        }
    }

    public static void main(String[] args) {
        ImpleDLL m = new ImpleDLL();
        m.append(new Node(1));
        m.append(new Node(2));
        m.append(new Node(3));
        m.push(new Node(10));
        m.printDll(head);
    }
}

File: Final_450/LinkedList/ImpleDll.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


def printDLL(head):
    temp = head
    print("forward direction")
    while(temp):
        print(temp.data, "<-->", end=" ")
        last = temp
        temp = temp.next
    print()
    print("backward direction")
    while(last):
        print(last.data, "<-->", end=" ")
        last = last.prev
    print()


def append(head, val):
    if(head == None):
        head = val
    temp = head
    while(temp.next != None):
        temp = temp.next
    temp.next = val
    val.prev = temp
    val.next = None


def insert(head, data):

    # inserts at the begining
    temp = Node(data)

    if (head == None):
        head = temp
    else:
        temp.next = head
        head.prev = temp
        head = temp
    return head


if __name__ == '__main__':

    head = None

    head = insert(head, 9)
    head = insert(head, 8)
    head = insert(head, 6)
    head = insert(head, 5)
    head = insert(head, 4)
    head = insert(head, 2)
    head = insert(head, 1)
    append(head, Node(11))
    printDLL(head)
